developer/src/kmc-kmn/src/kmw-compiler/javascript-strings.ts

```diff
@@ -247,20 +249,31 @@ export function JavaScript_Shift(fkp: KMX.KEY, FMnemonic: boolean): number {
   }
 
   if (fkp.ShiftFlags & KMX.KMXFile.ISVIRTUALKEY) {
-    if(IsKeyboardVersion10OrLater()) {
+    if(isKeyboardVersion10OrLater()) {
+      // don't attempt to upgrade to v10 at this point, only if we are already v10+
       // Full chiral modifier and state key support starts with KeymanWeb 10.0
       return fkp.ShiftFlags;
     }
 
     // Non-chiral support only and no support for state keys
     if (fkp.ShiftFlags & (KMX.KMXFile.LCTRLFLAG | KMX.KMXFile.RCTRLFLAG | KMX.KMXFile.LALTFLAG | KMX.KMXFile.RALTFLAG)) {   // I4118
+      // TODO: automatic version upgrade
+      // if(verifyMinimumRequiredKeymanVersion10()) {
+      //   // upgrade to v10 if possible
+      //   return fkp.ShiftFlags;
+      // }
       callbacks.reportMessage(KmwCompilerMessages.Warn_ExtendedShiftFlagsNotSupportedInKeymanWeb({line:fkp.Line, flags: 'LALT, RALT, LCTRL, RCTRL'}));
     }
 
     if (fkp.ShiftFlags & (
       KMX.KMXFile.CAPITALFLAG | KMX.KMXFile.NOTCAPITALFLAG | KMX.KMXFile.NUMLOCKFLAG | KMX.KMXFile.NOTNUMLOCKFLAG |
       KMX.KMXFile.SCROLLFLAG | KMX.KMXFile.NOTSCROLLFLAG)) {   // I4118
-      callbacks.reportMessage(KmwCompilerMessages.Warn_ExtendedShiftFlagsNotSupportedInKeymanWeb({line:fkp.Line, flags: 'CAPS and NCAPS'}));
+        // TODO: automatic version upgrade
+        // if(verifyMinimumRequiredKeymanVersion10()) {
+        // // upgrade to v10 if possible
+        // return fkp.ShiftFlags;
+        // }
+        callbacks.reportMessage(KmwCompilerMessages.Warn_ExtendedShiftFlagsNotSupportedInKeymanWeb({line:fkp.Line, flags: 'CAPS and NCAPS'}));
     }
```

Comment:
TODO: automatically upgrade version when chiral modifier or state flags are referenced in keyboard rules.